src/sequential/sequential.hpp

```diff
@@ -0,0 +1,198 @@
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// HELPER CLASSES
+////////////////////////////////////////////////////////////////////////////////
+
+enum SearchStatus {
+  FOUND_MATCH,
+  FOUND_SWAP,
+  FOUND_HOLE,
+  FOUND_NOHOLE,
+};
+
+/// @brief  Bucket for the Robin Hood hash table.
+struct SequentialBucket {
+  KeyType key;
+  ValueType value;
+  HashCodeType hashcode;
+  // A value of SIZE_MAX means that the bucket is empty. I do this hack so that
+  // we can fit this bucket into 4 words, which is more amenable to the hardware.
+  // A value of SIZE_MAX would be attrocious for performance anyways.
+  size_t offset;
+
+  SequentialBucket(KeyType key, ValueType value, HashCodeType hashcode, size_t offset) {
+    LOG_TRACE("Enter");
+    this->key = key;
+    this->value = value;
+    this->hashcode = hashcode;
+    this->offset = offset;
+  }
+
+  SequentialBucket() {
+    LOG_TRACE("Enter");
+    this->key = 0;
+    this->value = 0;
+    this->hashcode = 0;
+    this->offset = SIZE_MAX;
+  }
+
+  bool is_empty();
+
+  void set(const KeyType key,
+           const ValueType value,
+           const HashCodeType hashcode,
+           const size_t offset);
+
+  void swap(SequentialBucket &other);
+
+  bool equal_key(const KeyType key, const HashCodeType hashcode);
+
+  /// @brief  Pretty print bucket
+  ///
+  /// This prints in the format (using Python's f-string syntax):
+  /// f"({hashcode}=>{home}+{offset}) {key}: {value}"
+  void print(const size_t size);
+};
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// STATIC HELPER FUNCTIONS
+////////////////////////////////////////////////////////////////////////////////
+
+/// Source: https://stackoverflow.com/questions/664014/what-integer-hash-function-are-good-that-accepts-an-integer-hash-key
+HashCodeType
+hash(const KeyType key);
+
+size_t
+get_home(const HashCodeType hashcode, const size_t size);
+
+/// @brief  Get real bucket index.
+size_t
+get_probe(const size_t home, const size_t offset, const size_t size);
+
+/// @brief  Get offset from home or where it would be if not found.
+///         Return SIZE_MAX if no hole is found.
+std::pair<SearchStatus, size_t>
+get_wouldbe_offset(      std::vector<SequentialBucket> &tmp_buckets,
+                   const KeyType key,
+                   const HashCodeType hashcode,
+                   const size_t home);
+
+/// @brief  Insert but assume no resize is necessary.
+int
+insert_without_resize(      std::vector<SequentialBucket> &tmp_buckets,
+                          const KeyType key,
+                          const ValueType value,
+                          const HashCodeType hashcode);
+
+////////////////////////////////////////////////////////////////////////////////
+/// HASH TABLE CLASS
+////////////////////////////////////////////////////////////////////////////////
+
+class SequentialRobinHoodHashTable {
+private:
```

Comment:
I personally prefer to have public members come first so that anyone browsing the code sees the public-facing API first before the implementation details.